abci/client/committing_client.go

```diff
@@ -0,0 +1,361 @@
+package abcicli
+
+import (
+	"context"
+
+	types "github.com/tendermint/tendermint/abci/types"
+	tmsync "github.com/tendermint/tendermint/internal/libs/sync"
+	"github.com/tendermint/tendermint/libs/service"
+)
+
+var _ Client = (*committingClient)(nil)
+
+// NOTE: use defer to unlock mutex because Application might panic (e.g., in
+// case of malicious tx or query). It only makes sense for publicly exposed
+// methods like CheckTx (/broadcast_tx_* RPC endpoint) or Query (/abci_query
+// RPC endpoint), but defers are used everywhere for the sake of consistency.
+type committingClient struct {
+	service.BaseService
```

Comment:
forgive my ignorance, but I'm not sure how/why this needs to compose the base service methods, and I think in order for the client to be (usable as) a `service.Service` if you implement some additional methods.